### PR TITLE
MultiPop: Accessor and replacement methods

### PR DESCRIPTION
**Pull request by darizasu**
This PR improves the usability of the `MultiPop` class by adding convenient accessor and replacement methods, plus utilities for working with named and/or nested `MultiPop` objects. It also includes small improvements to existing MultiPop merge helpers.

**Summary of changes**

- `MultiPop` now supports `$` and `names()` access, plus replacement methods for `[<-`, `[[<-`, `$<-`, and `names<-`, making it easier to build, edit, and name `MultiPop` objects programmatically.
- The `show()` method for `MultiPop` now displays item indices, names (when present), and nested structure in a more informative way (with tests).
- Added `unnameMultiPop()` to remove names from a `MultiPop` at specific nesting levels (or at all levels), with input validation and tests.
- Improved merge/flatten behavior and tests (e.g., nested `MultiPop`s supported by `mergePops()`, additional validation via `validObject()`).

**Commit overview**

- c35587ce: `mergePops()` now correctly handles nested `MultiPop` objects; tests updated accordingly, fixing the problem described in #253.
- 237b05c8: Documentation + tests improvements for `flattenMultiPop()` / `mergeMultiPops()`, validate returned `MultiPop` objects, and expand merge tests (including nested inputs and `NULL` handling).
- 0a1b8859: `newPop()` can now create an empty `Pop` when `rawPop` is missing and `ploidy` is provided; docs updated.
- 645f0d90: Add `MultiPop` accessors/replacement methods (`$`, `names`, `[<-`, `[[<-`, `$<-`, `names<-`) and comprehensive tests; update NAMESPACE imports.
- 68baaa80: Improve `MultiPop` `show()` output to include names and handle empty / missing names cleanly; add `show()` tests.
- 346ce97b: Add exported `unnameMultiPop()` (plus internal helpers) and tests.
- 448bbde8: Update NEWS to reflect new `MultiPop` functionality.

**Notes / questions for review**

- These changes are intended to be backwards-compatible: Existing `MultiPop` construction and indexing via `[` / `[[` continues to work, while the new methods enable more ergonomic workflows (especially for named/nested `MultiPop`s).
- The accessor and replacement methods are designed to follow base R list semantics where possible:
  - Accessing non-existent elements by name or index (via `$` / `[[`) returns `NULL`.
  - Assigning `NULL` via a replacement method removes the element.
  - Assigning to a new name/index appends a new element to the `MultiPop`.
  - Subsetting with `[` can return an empty `MultiPop` when no elements are selected.
- `unnameMultiPop()` currently uses `level = Inf` as a "wildcard" to remove names at every level in a nested MultiPop. If you prefer a different convention (e.g., a separate argument such as `all = TRUE`, or mapping to `force = TRUE`), I’m happy to adjust.

_**Side note**_: `unnameMultiPop()` is a standalone function (and not an `unname()` method @gregorgorjanc)

I considered adding an S4 method for `base::unname()` so that `unname(mp)` would behave like `unnameMultiPop(mp, level = ...)`. The main difficulties are:

- `base::unname()` has the fixed signature `unname(obj, force = FALSE)`, so any `MultiPop`-specific argument like `level` would need to be hidden in `...`.
- The `force` argument would be ignored for `MultiPop` (unless we repurpose it, e.g. `force = TRUE` meaning “remove at all levels”).
- The fixed signature `obj` and `fixed` would also require documentation in the MultiPop definition. This may be confusing for an effectively unused argument (`force`), and overall felt more confusing than providing an explicit `unnameMultiPop()` function.

For these reasons, I kept `unnameMultiPop()` as the user-facing function, but I’m open to thoughts about switching to an `unname()` method if that’s preferred.

### CODE REVIEW

File: R/Class-Pop.R [hunk 1]
```diff
@@ -587,10 +587,14 @@ setMethod("length",
 #' \code{\link{MapPop-class}} or \code{\link{NamedMapPop-class}}.
 #' The function is intended for use with output from functions such
 #' as \code{\link{runMacs}}, \code{\link{newMapPop}}, or
-#' \code{\link{quickHaplo}}.
+#' \code{\link{quickHaplo}}. However, it can also be used to create an
```
Comment (gregorgorjanc) [inline R/Class-Pop.R]:
```suggestion
#' \code{\link{quickHaplo}}. It can also be used to create an
```

Comment (gregorgorjanc) [reply R/Class-Pop.R]: @gaynorr we added this so we are consistent across the codebase.

File: R/Class-Pop.R [hunk 2]
```diff
@@ -1108,6 +1148,150 @@ setMethod("[[",
...
+
+  # List semantics handles numeric/character/logical indices, recycling, names, expansion
+  x@pops[i] = value
+  validObject(x)
+  return(x)
+})
+
+#' @describeIn MultiPop Replace contents of a single element in MultiPop
+setReplaceMethod("[[", signature(x = "MultiPop"), function(x, i, value) {
+  stopifnot("index required" = !missing(i))
```
Comment (gregorgorjanc) [inline R/Class-Pop.R]: I find this code a bit cryptic and think that best to strive for simple/clear code is important. Or maybe I am just now used to this form of using `stopifnot()`?

File: R/Class-Pop.R [hunk 3]
```diff
@@ -1204,3 +1388,138 @@ isMultiPop = function(x) {
...
+#'   If \code{level = c(a, b)}, names are removed only at levels \code{a}
+#'   and \code{b}. Levels must be >= 1; an error is raised if any
+#'   requested level is deeper than the MultiPop's maximum depth.
+#'   The top level is \code{1}.
+#'
+#' @return A \code{\link{MultiPop-class}} object with names removed at the
+#'   requested levels.
+#'
+#' @details
+#' - Top-level elements of a MultiPop have level 1; nested MultiPops increase
```
Comment (gregorgorjanc) [inline R/Class-Pop.R:1410]: Does this render in man page as list items or is it all wrapped into one paragraph? Check.

Comment (gregorgorjanc) [reply R/Class-Pop.R:1410]: Aha, is this what `\cr` are for?

File: R/mergePops.R [hunk 4]
```diff
@@ -374,14 +378,8 @@ mergeMultiPops = function(..., level=0){
   popList = list(...)
   classes = do.call("c", lapply(popList, class))
 
-  if(any(classes == "NULL")){
```
Comment (gregorgorjanc) [inline R/mergePops.R:377]: @darizasu was this redundant piece of code hence why the removal, but what if we get an object whose class is NULL - hmm, is that even possible (suggesting that indeed this should be removed)?

File: tests/testthat/test-MultiPop.R [hunk 5]
```diff
@@ -0,0 +1,314 @@
+
+test_that("MultiPop assignment and replace methods", {
+
+  founderPop = quickHaplo(nInd=100, nChr=1, segSites=10)
```
Comment (gregorgorjanc) [inline tests/testthat/test-MultiPop.R]: Do we need 100 ind? Just trying to keep tests small as we are adding more and more and checking time can pile up;)

File: tests/testthat/test-MultiPop.R [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+  SP = SimParam$new(founderPop)
+  SP$addTraitA(10, mean = c(0, 0), var = c(1, 1))
```
Comment (gregorgorjanc) [inline tests/testthat/test-MultiPop.R:7]: I think we are encouraged to use single threads so set SP$nThreads = 1L. do that across all your tests suites (for each `test_that()` block).

File: tests/testthat/test-MultiPop.R [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+  # Append elements to a MultiPop using `[[<-` and defined names
+  mpA = newMultiPop()
+  mpA[["Group1"]] = pop[1:5]
+  mpA[["Group2"]] = pop[6:10]
+  expect_identical(names(mpA), c("Group1", "Group2"))
+  expect_identical(mpA[["Group1"]], pop[1:5])
+
+  # Append elements to a MultiPop using `[<-` and defined names
+  mpB = newEmptyMultiPop()
+  mpB[c('Group1','Group2')] = list(pop[1:5], pop[6:10])
```
Comment (gregorgorjanc) [inline tests/testthat/test-MultiPop.R]: Should `list()` be `newMultiPop()` in this line?

File: tests/testthat/test-MultiPop.R [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+
+  # Test that the MultiPops created with different methods are identical
+  expect_identical(mpA, mpB)
+  expect_identical(mpA, mpC)
+  expect_identical(mpB, mpC)
+
+  # Replace elements in a MultiPop using `[[<-`, `[<-`, and `$<-`
+  mpA[["Group1"]] = pop[11:15]
+  mpA[[2]] = pop[16:20]
+  mpB[c('Group1')] = list(pop[11:15])
```
Comment (gregorgorjanc) [inline tests/testthat/test-MultiPop.R]: As before.

File: tests/testthat/test-MultiPop.R (hunk 7 above)
Comment (gregorgorjanc) [reply tests/testthat/test-MultiPop.R]: I know they are effectively identical, but that's implementation detail that we don't want to rely on in testing. The implementation will not likely change, but best to be consistent with high-level API rather than mix that with low-level implementation.

File: tests/testthat/test-MultiPop.R (hunk 8 above)
Comment (gregorgorjanc) [reply tests/testthat/test-MultiPop.R]: I see some more below so consider changing them.

File: tests/testthat/test-MultiPop.R (hunk 5 above)
Comment (gregorgorjanc) [reply tests/testthat/test-MultiPop.R]: Can any of your other tests be made smaller too?

File: tests/testthat/test-MultiPop.R [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+  names(mpD) = NULL
+  expect_null(names(mpD))
+
+  # Multipop with 3 levels of nesting and direct name assignment
+  mp2 = newMultiPop(pop1 = pop[1:20], pop2 = pop[21:40],
+                    mpA = newMultiPop(pop3 = pop[41:60],
+                                      mpB = newMultiPop(pop4 = pop[61:80], pop5 = pop[81:100])))
+  mp2copy = mp2
+
+  # `names` returns names of the top-level elements in a MultiPop, even if it has nested structure
```
Comment (gregorgorjanc) [inline tests/testthat/test-MultiPop.R:87]: since we now have `unnameMultiPop` should we ever wanted `names` to work across  the nested structure, we could just write `namesMultiPop`;)

File: tests/testthat/test-MultiPop.R [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+
+  expect_error(
+    dollar_set(mp, "A", 1L),
+    "value must be a Pop or MultiPop",
+    fixed = TRUE
+  )
+})
+
+test_that("MultiPop show", {
+  founderPop <- quickHaplo(nInd = 40, nChr = 1, segSites = 10)
```
Comment (gregorgorjanc) [inline tests/testthat/test-MultiPop.R]: Again reduce as much as possible. thanks!

File: tests/testthat/test-MultiPop.R [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+
+  # Name suppression branch for "" and NA
+  names(mp) <- c("", NA_character_)
+  out2 <- capture.output(show(mp))
+  expect_false(any(grepl('"NA" - ', out2, fixed = TRUE)))
+  expect_false(any(grepl('"" - ', out2, fixed = TRUE)))
+})
+
+test_that("Multipop unnameMultiPop", {
+  founderPop <- quickHaplo(nInd = 100, nChr = 1, segSites = 10)
```
Comment (gregorgorjanc) [inline tests/testthat/test-MultiPop.R]: Again;)

**Review by gregorgorjanc — COMMENTED** [review]
@darizasu I went over this. It looks very good! There are some minor bits I suggest you address - see specific code/doc comments.

File: tests/testthat/test-MultiPop.R (hunk 7 above)
Comment (darizasu) [reply tests/testthat/test-MultiPop.R]: The assignment method currently supports both `list(pop)` and `newMultiPop(pop)`. As briefly discussed, I'll keep both in the tests. Done in 7751e0f1cb8b2ad2e3a6990288409bded72d48ff

File: R/Class-Pop.R (hunk 3 above)
Comment (darizasu) [reply R/Class-Pop.R:1410]: Yes, exactly. I found the `\cr` quite handy to render bullet points in the 'Description' section.

File: R/Class-Pop.R (hunk 2 above)
Comment (darizasu) [reply R/Class-Pop.R]: I agree, it works but it's confusing when you read it. I changed it to `if (missing(i)) stop("index required")` in 7751e0f1cb8b2ad2e3a6990288409bded72d48ff

File: tests/testthat/test-MultiPop.R (hunk 5 above)
Comment (darizasu) [reply tests/testthat/test-MultiPop.R]: No, we don't need so many. I reduced it to up to 10 individuals in 7751e0f1cb8b2ad2e3a6990288409bded72d48ff

File: tests/testthat/test-MultiPop.R (hunk 9 above)
Comment (darizasu) [reply tests/testthat/test-MultiPop.R:87]: That makes sense. As discussed, I'll implement it later if someone asks for it.

File: R/mergePops.R (hunk 4 above)
Comment (darizasu) [reply R/mergePops.R:377]: Yes, I thought it was redundant with `mergePops()` so I removed it. However, I realized it can be useful in edge cases, so I brought it back in 7751e0f1cb8b2ad2e3a6990288409bded72d48ff.

File: tests/testthat/test-MultiPop.R (hunk 7 above)
Comment (gregorgorjanc) [reply tests/testthat/test-MultiPop.R]: @darizasu do we mention anywhere in the docs that both `list` and `MultiPop` classes are supported for some/all? operations?

**Comment by darizasu** [comment]
Thank you for your feedback @gregorgorjanc!  I went through your comments and applied your suggestions in 7751e0f1cb8b2ad2e3a6990288409bded72d48ff. I'll mark this as ready for review for @gaynorr
